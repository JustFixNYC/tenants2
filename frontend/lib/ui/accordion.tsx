import React from "react";
import { StaticImage } from "./static-image";

const ChevronIcon = () => (
  <StaticImage ratio="is-16x16" src="frontend/img/chevron.svg" alt="" />
);

export const Accordion = (props: {
  question: string;
  children: React.ReactNode;
  questionClassName?: string;
  extraClassName?: string;
  /** When set to true, the accordion is open on initial view. */
  isExpanded?: boolean;
}) => {
  const extraClassName = props.extraClassName ?? "jf-space-below-2rem";

  return (
    <div className={`jf-accordion-item ${extraClassName}`}>
      <details
        className={`has-text-left ${extraClassName}`}
        open={props.isExpanded}
      >
        <summary>
          <div className="media">
            <div className="media-content">
              <span
                className={
                  props.questionClassName ??
                  "is-size-6 has-text-primary jf-has-text-underline"
                }
              >
                {props.question}
              </span>
            </div>
            <div className="media-right">
              <ChevronIcon />
            </div>
          </div>
        </summary>
        {props.children}
      </details>
    </div>
  );
};
