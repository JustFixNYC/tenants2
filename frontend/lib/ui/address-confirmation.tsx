import React, { Children } from "react";
import { BackOrUpOneDirLevel, Modal } from "./modal";
import { Link } from "react-router-dom";
import {
  isBoroughChoice,
  getBoroughChoiceLabels,
} from "../../../common-data/borough-choices";
import { CenteredButtons } from "./centered-buttons";

type AddressAndBorough = {
  /** A NYC street name and number, e.g. "150 court st". */
  address: string;
  /** A NYC borough choice, e.g. "STATEN_ISLAND". */
  borough: string;
};

export type ConfirmAddressModalProps = AddressAndBorough & {
  /**
   * The route to the next step of the user flow, if the user confirms the
   * correctness of the address.
   */
  nextStep: string;
};

/**
 * A modal that we present the user if the address they entered is different from
 * the one we geocoded on the server. They are given the option to go back to the
 * previous step (to change their address) or to continue to the next step.
 */
export function ConfirmAddressModal(
  props: ConfirmAddressModalProps
): JSX.Element {
  let borough = "";

  if (isBoroughChoice(props.borough)) {
    borough = getBoroughChoiceLabels()[props.borough];
  }

  return (
    <Modal
      title="Is this your address?"
      withHeading
      onCloseGoTo={BackOrUpOneDirLevel}
      render={(ctx) => (
        <>
          <p>
            {props.address}, {borough}
          </p>
          <CenteredButtons>
            <Link to={props.nextStep} className="button is-primary is-medium">
              Yes!
            </Link>
            <Link {...ctx.getLinkCloseProps()} className="button is-text">
              No, go back.
            </Link>
          </CenteredButtons>
        </>
      )}
    />
  );
}

export type RedirectToAddressConfirmationOrNextStepOptions = {
  /** The address the user input. */
  input: AddressAndBorough;
  /** The resolved address the server geocoded, based on the user's input. */
  resolved: AddressAndBorough;
  /** The route to go to if the user needs to confirm the correctness of the address. */
  confirmation: string;
  /**
   * The route to go to if the user doesn't need to confirm the correctness of the address
   * (i.e., if the resolved address is semantically equivalent to the input address).
   */
  nextStep: string;
};

/**
 * Given the user's input address and the resolved address that the server geocoded, either
 * present the user with a confirmation modal or send them on to the next step in their flow.
 *
 * Returns the route to redirect the user to.
 */
export function redirectToAddressConfirmationOrNextStep(
  options: RedirectToAddressConfirmationOrNextStepOptions
): string {
  const { input, resolved } = options;
  if (
    areAddressesTheSame(resolved.address, input.address) &&
    resolved.borough === input.borough
  ) {
    return options.nextStep;
  }
  return options.confirmation;
}

/** Returns whether the given street name and numbers are semantically equivalent. */
export function areAddressesTheSame(a: string, b: string): boolean {
  return a.trim().toUpperCase() === b.trim().toUpperCase();
}
