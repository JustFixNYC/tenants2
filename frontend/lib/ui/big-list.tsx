import React from "react";
import classnames from "classnames";

export type BigListProps = {
  /** The list items. */
  children: JSX.Element[];

  /** The ol list element will be given this class, if provided. */
  listClassName?: string;

  /** The content of each list item will be given this class, if provided. */
  itemClassName?: string;
};

/**
 * An ordered list with very big numbers.
 *
 * Each child should be a `<li>` without any props other than children and,
 * optionally, a `key`.
 */
export function BigList(props: BigListProps) {
  return (
    <ol className={classnames("jf-biglist", props.listClassName)}>
      {React.Children.map(props.children, (child, i) => (
        <li key={child.key === null ? i : child.key}>
          <div className="jf-biglist-counter" />
          <div className={props.itemClassName}>{child.props.children}</div>
        </li>
      ))}
    </ol>
  );
}
