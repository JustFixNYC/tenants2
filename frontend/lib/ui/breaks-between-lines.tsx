import React from "react";

export const BreaksBetweenLines: React.FC<{ lines: string | string[] }> = (
  props
) => {
  const strLines =
    typeof props.lines === "string" ? props.lines.split("\n") : props.lines;
  const lines: Array<string | JSX.Element> = [];

  for (let i = 0; i < strLines.length; i++) {
    lines.push(strLines[i]);
    if (i < strLines.length - 1) {
      lines.push(<br key={i} />);
    }
  }

  return <>{lines}</>;
};
