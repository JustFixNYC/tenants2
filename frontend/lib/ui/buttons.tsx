import React from "react";
import classnames from "classnames";

import { bulmaClasses, BulmaClassName } from "./bulma";
import { Link, LinkProps } from "react-router-dom";
import { LocationDescriptor } from "history";

type ProgressButtonsOptions =
  | {
      children: JSX.Element[];
    }
  | {
      children?: undefined;
      back: string;
      isLoading: boolean;
      nextLabel?: string;
    };

/**
 * A component that can be used in two different ways:
 *
 *   1. As a container for two child buttons of any type.
 *   2. As a childless component with props that will automatically
 *      render back/next buttons with the most common options.
 */
export function ProgressButtons(props: ProgressButtonsOptions) {
  return (
    <div className="buttons jf-two-buttons">
      {"children" in props ? (
        props.children
      ) : (
        <>
          <BackButton to={props.back} />
          <NextButton isLoading={props.isLoading} label={props.nextLabel} />
        </>
      )}
    </div>
  );
}

/**
 * A component that renders back/next buttons that function as internal links,
 * as in they don't involve any form submission or mutation. 
 */
export function ProgressButtonsAsLinks(props: {
  back: LocationDescriptor<any>;
  backLabel?: string;
  next: LocationDescriptor<any>;
  nextLabel?: string;
}): JSX.Element {
  return (
    <div className="buttons jf-two-buttons">
      <BackButton to={props.back} label={props.backLabel} />
      <Link
        to={props.next}
        className="button is-primary is-medium jf-is-next-button"
      >
        {props.nextLabel || "Next"}
      </Link>
    </div>
  );
}

/** A back button, meant to go back to the previous step in a flow. */
export function BackButton(props: {
  buttonClass?: BulmaClassName;
  to: LocationDescriptor<any>;
  label?: string;
}): JSX.Element {
  return (
    <Link
      to={props.to}
      className={
        "jf-is-back-button " +
        bulmaClasses("button", props.buttonClass || "is-light", "is-medium")
      }
    >
      {props.label || "Back"}
    </Link>
  );
}

/** A next button that submits a form and takes the user to the next step in a flow. */
export function NextButton(props: {
  buttonClass?: BulmaClassName;
  buttonSizeClass?: BulmaClassName;
  isFullWidth?: boolean;
  isLoading: boolean;
  label?: string;
}): JSX.Element {
  return (
    <button
      type="submit"
      className={
        "jf-is-next-button " +
        bulmaClasses(
          "button",
          props.buttonClass || "is-primary",
          props.buttonSizeClass || "is-medium",
          {
            "is-loading": props.isLoading,
            "is-fullwidth": props.isFullWidth,
          }
        )
      }
    >
      {props.label || "Next"}
    </button>
  );
}

export function CenteredPrimaryButtonLink(props: LinkProps): JSX.Element {
  props = {
    ...props,
    className: classnames(
      props.className,
      bulmaClasses("button", "is-primary", "is-large"),
      "jf-is-extra-wide"
    ),
  };
  return (
    <p className="has-text-centered">
      <Link {...props} />
    </p>
  );
}
