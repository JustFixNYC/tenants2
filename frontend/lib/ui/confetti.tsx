import React from "react";

import { confetti } from "../../vendor/confetti";
import { assertNotNull } from "@justfixnyc/util";
import { createPortal } from "react-dom";
import autobind from "autobind-decorator";

export const CONFETTI_WRAPPER_CLASS = "jf-confetti-wrapper";

/**
 * Return whether the browser supports the 'pointer-events' CSS
 * property. It's critical to making it possible for users to
 * actually click through the canvas and into the document behind
 * it.
 */
function supportsPointerEvents(): boolean {
  const html = document.documentElement;
  return html ? "pointerEvents" in html.style : false;
}

export function ensurePointerEventsIsNone(el: Element) {
  if (window.getComputedStyle(el).pointerEvents !== "none") {
    throw new Error('pointer-events of element is not "none"!');
  }
}

export interface ConfettiProps extends ConfettiCanvasProps {}

interface ConfettiState {
  mounted: boolean;
}

export default class Confetti extends React.Component<
  ConfettiProps,
  ConfettiState
> {
  container: HTMLDivElement | null;

  constructor(props: ConfettiProps) {
    super(props);
    this.container = null;
    this.state = {
      mounted: false,
    };
  }

  componentDidMount() {
    this.container = document.createElement("div");
    this.container.className = CONFETTI_WRAPPER_CLASS;
    document.body.appendChild(this.container);
    this.setState({ mounted: true });
  }

  componentWillUnmount() {
    document.body.removeChild(assertNotNull(this.container));
  }

  render() {
    if (!this.state.mounted || !supportsPointerEvents()) {
      return null;
    }

    return createPortal(
      <ConfettiCanvas {...this.props} />,
      assertNotNull(this.container)
    );
  }
}

interface ConfettiCanvasProps {
  /**
   * Number of seconds during which any confetti that falls to
   * the bottom of the screen will be "regenerated" and moved
   * above the screen so it will fall once again. Defaults to 0,
   * which means it confetti will regenerate forever.
   */
  regenerateForSecs?: number;
}

interface ConfettiCanvasState {
  hasFinished: boolean;
}

class ConfettiCanvas extends React.Component<
  ConfettiCanvasProps,
  ConfettiCanvasState
> {
  canvasRef: React.RefObject<HTMLCanvasElement> = React.createRef();
  confettiCtx: any = null;

  constructor(props: ConfettiCanvasProps) {
    super(props);
    this.state = {
      hasFinished: false,
    };
  }

  @autobind
  handleResize() {
    this.confettiCtx.resize();
  }

  @autobind
  handleConfettiFinished() {
    this.cleanupConfetti();
    this.setState({ hasFinished: true });
  }

  cleanupConfetti() {
    if (this.confettiCtx !== null) {
      this.confettiCtx.stop();
      window.removeEventListener("resize", this.handleResize, false);
      this.confettiCtx = null;
    }
  }

  componentDidMount() {
    const canvas = assertNotNull(this.canvasRef.current);
    ensurePointerEventsIsNone(canvas);
    this.confettiCtx = new confetti.Context(
      canvas,
      this.props.regenerateForSecs,
      this.handleConfettiFinished
    );
    this.confettiCtx.start();
    window.addEventListener("resize", this.handleResize, false);
  }

  componentWillUnmount() {
    this.cleanupConfetti();
  }

  render() {
    return this.state.hasFinished ? null : <canvas ref={this.canvasRef} />;
  }
}
