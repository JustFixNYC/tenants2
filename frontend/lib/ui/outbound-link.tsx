import React from "react";
import { DetailedHTMLProps, AnchorHTMLAttributes, MouseEvent } from "react";
import hardRedirect from "../browser-redirect";
import { callOnceWithinMs } from "../util/util";
import { ga } from "../analytics/google-analytics";
import { logAmplitudeOutboundLinkClick } from "../analytics/amplitude";

/**
 * When we call ga(), it's actually possible that GA never loaded,
 * and window.ga is just the stub in the GA snippet that came with
 * our page. There are other potential edge cases where our
 * hit callback might never be called, so let's ensure that we
 * navigate regardless once the given number of milliseconds
 * have elapsed, just to be safe.
 */
const OUTBOUND_LINK_TIMEOUT_MS = 1000;

/** An event handler executed when a user clicks on an outbound link. */
export function handleOutboundLinkClick(e: MouseEvent<HTMLAnchorElement>) {
  // If any modifier key is pressed, odds are that the user is trying to
  // invoke some browser-specific behavior to e.g. open the link in a
  // new window or tab. We don't want to break that.
  const isModifierPressed = e.altKey || e.ctrlKey || e.metaKey || e.shiftKey;
  const { href, target } = e.currentTarget;
  const willOpenInNewWindow = target && target !== window.name;
  if (!isModifierPressed && !willOpenInNewWindow) {
    ga("send", "event", "outbound", "click", href, {
      transport: "beacon",
      hitCallback: callOnceWithinMs(
        () => hardRedirect(href),
        OUTBOUND_LINK_TIMEOUT_MS
      ),
    });
    // Note that we're not currently logging Amplitude events in this branch,
    // but that should be OK because we almost always follow the other branch
    // in practice (virtually all our external links open in a new window).
    e.preventDefault();
  } else {
    ga("send", "event", "outbound", "click", href);
    logAmplitudeOutboundLinkClick(href);
  }
}

export type OutboundLinkProps = DetailedHTMLProps<
  AnchorHTMLAttributes<HTMLAnchorElement>,
  HTMLAnchorElement
> & {
  /** The "href" prop is required on outbound links, not optional. */
  href: string;
};

const defaultOutboundLinkProps = {
  target: "_blank",
  rel: "noopener noreferrer",
};

/**
 * A react component that encapsulates a link to an external website,
 * which we want to track with analytics.
 */
export function OutboundLink(props: OutboundLinkProps): JSX.Element {
  props = { ...defaultOutboundLinkProps, ...props };
  const { onClick, ...otherProps } = props;
  return (
    <a
      {...otherProps}
      onClick={(e) => {
        handleOutboundLinkClick(e);
        if (onClick) onClick(e);
      }}
    >
      {props.children}
    </a>
  );
}
