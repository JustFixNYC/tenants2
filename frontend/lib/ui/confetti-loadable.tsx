import React from "react";
import { ConfettiProps } from "./confetti";
import { SimpleProgressiveEnhancement } from "./progressive-enhancement";
import loadable from "@loadable/component";

const LoadableConfetti = loadable(() => import("./confetti"), {
  // We don't want to display anything while the confetti is loading.
  fallback: <></>,
  ssr: false,
});

export function ProgressiveLoadableConfetti(props: ConfettiProps): JSX.Element {
  return (
    <SimpleProgressiveEnhancement>
      <LoadableConfetti {...props} />
    </SimpleProgressiveEnhancement>
  );
}
