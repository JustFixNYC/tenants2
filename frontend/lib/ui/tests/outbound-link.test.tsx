import React from "react";
import { handleOutboundLinkClick, OutboundLink } from "../../ui/outbound-link";
import ReactTestingLibraryPal from "../../tests/rtl-pal";
import { setHardRedirector } from "../../browser-redirect";

describe("OutboundLink", () => {
  const gaMock = jest.fn();
  const hardRedirect = jest.fn();
  const callHitCallback = () => {
    gaMock.mock.calls[0][5].hitCallback();
  };

  beforeEach(() => {
    jest.useFakeTimers();
    window.ga = gaMock;
    setHardRedirector(hardRedirect);
    gaMock.mockReset();
    hardRedirect.mockReset();
  });

  it("sends hit to GA on click and then redirects", () => {
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink href="https://boop.com/" target={undefined}>
          boop
        </OutboundLink>
      )
    );
    const wasDefaultPrevented = !pal.clickButtonOrLink("boop");
    expect(gaMock.mock.calls).toHaveLength(1);
    expect(gaMock.mock.calls[0].slice(0, 5)).toEqual([
      "send",
      "event",
      "outbound",
      "click",
      "https://boop.com/",
    ]);
    expect(wasDefaultPrevented).toBe(true);

    jest.advanceTimersByTime(400);
    expect(hardRedirect.mock.calls).toHaveLength(0);
    callHitCallback();
    expect(hardRedirect.mock.calls).toEqual([["https://boop.com/"]]);
  });

  it('does not prevent default if "target" prop is set', () => {
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink href="https://bap.com/" target="_blank">
          bap
        </OutboundLink>
      )
    );
    const wasDefaultPrevented = !pal.clickButtonOrLink("bap");
    expect(gaMock.mock.calls).toEqual([
      ["send", "event", "outbound", "click", "https://bap.com/"],
    ]);
    expect(hardRedirect.mock.calls).toHaveLength(0);
    expect(wasDefaultPrevented).toBe(false);
  });

  it("redirects after a timeout if GA has not responded", () => {
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink href="https://blorp.com/" target={undefined}>
          blorp
        </OutboundLink>
      )
    );
    pal.clickButtonOrLink("blorp");
    jest.advanceTimersByTime(1001);
    expect(hardRedirect.mock.calls).toEqual([["https://blorp.com/"]]);
  });

  it("prevents default click behavior when using GA", () => {
    const preventDefault = jest.fn();
    const e = { currentTarget: { href: "blah" }, preventDefault } as any;
    handleOutboundLinkClick(e);
    expect(gaMock.mock.calls).toHaveLength(1);
    expect(preventDefault.mock.calls).toHaveLength(1);
  });

  it("sends GA event but does not prevent default when a modifier key is pressed", () => {
    const preventDefault = jest.fn();
    const e = {
      shiftKey: true,
      preventDefault,
      currentTarget: { href: "http://boop" },
    } as any;
    handleOutboundLinkClick(e);
    expect(gaMock.mock.calls).toEqual([
      ["send", "event", "outbound", "click", "http://boop"],
    ]);
    expect(preventDefault.mock.calls).toHaveLength(0);
  });

  it("calls onClick prop if passed", () => {
    const onClick = jest.fn();
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink href="https://boop.com/" onClick={onClick}>
          boop
        </OutboundLink>
      )
    );
    pal.clickButtonOrLink("boop");
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('Sets default values for "target" and "rel" props', () => {
    const pal = new ReactTestingLibraryPal(
      <OutboundLink href="https://bap.com/">bap</OutboundLink>
    );
    expect(pal.rr.container).toMatchSnapshot();
  });
  it("Doesn't override user value for 'target' but still inserts default 'rel' prop", () => {
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink href="https://bap.com/" target="boop">
          bap
        </OutboundLink>
      )
    );
    expect(pal.rr.container).toMatchSnapshot();
  });
  it('Allows "target" and "rel" props to be set to empty strings', () => {
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink href="https://bap.com/" target="" rel="">
          bap
        </OutboundLink>
      )
    );
    expect(pal.rr.container).toMatchSnapshot();
  });
  it('Allows "target" and "rel" props to be set to undefined', () => {
    const pal = new ReactTestingLibraryPal(
      (
        <OutboundLink
          href="https://bap.com/"
          target={undefined}
          rel={undefined}
        >
          bap
        </OutboundLink>
      )
    );
    expect(pal.rr.container).toMatchSnapshot();
  });
});
