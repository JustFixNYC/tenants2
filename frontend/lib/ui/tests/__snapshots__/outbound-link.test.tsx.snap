// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutboundLink Allows "target" and "rel" props to be set to empty strings 1`] = `
<div>
  <a
    href="https://bap.com/"
    rel=""
    target=""
  >
    bap
  </a>
</div>
`;

exports[`OutboundLink Allows "target" and "rel" props to be set to undefined 1`] = `
<div>
  <a
    href="https://bap.com/"
  >
    bap
  </a>
</div>
`;

exports[`OutboundLink Doesn't override user value for 'target' but still inserts default 'rel' prop 1`] = `
<div>
  <a
    href="https://bap.com/"
    rel="noopener noreferrer"
    target="boop"
  >
    bap
  </a>
</div>
`;

exports[`OutboundLink Sets default values for "target" and "rel" props 1`] = `
<div>
  <a
    href="https://bap.com/"
    rel="noopener noreferrer"
    target="_blank"
  >
    bap
  </a>
</div>
`;
